### PR TITLE
[Major]: refactor: use`RouterService` instead of Ember `Router` 

### LINKED ISSUE #255: Use `RouterService` instead of extending Ember `Router`
Similar to https://github.com/ember-app-scheduler/ember-app-scheduler/issues/772.

Looking at what actually happens in [`EmberRouterScroll`](https://github.com/DockYard/ember-router-scroll/blob/df23a157d0ef5ac306a28664a6e0411ad2191bc1/addon/index.js#L92) it only seems like `routeWillChange` and `routeDidChange` are bound. These events are available on the [`RouterService`](https://api.emberjs.com/ember/release/classes/RouterService) as well. To me it looks like we could now easily refactor this addon to avoid the longer inheritance chain.

### PR DESCRIPTION
**Pull request by snewcomer**
close #256 close #255

This will be released as a major bump -> `4.0`

- [x] remove `locations/router-scroll.js`. This has been deprecated and you should rely on Ember's own implementation instead of extending from us.  #193 
- [x] Remove `EmberRouterScroll` that you used to import and extend in your application.  We handle that for you now and eagerly initialize the service for you.  We did this because we can rely on `RouterService` events instead of [this](https://github.com/DockYard/ember-router-scroll/blob/07e97a496bf5981c8a74d1607625ff9e144e3956/addon/index.js#L68).

### CODE REVIEW

**Comment by snewcomer** [comment]
@buschtoens 👋 Let me know what you think!!

**Comment by snewcomer** [comment]
@simonihmig as well if you have a moment!

File: addon/services/router-scroll.js [hunk 1]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+    if (url && url.indexOf('#') > -1 && hashElement) {
+      scrollPosition = { x: hashElement.offsetLeft, y: hashElement.offsetTop };
+    } else {
+      scrollPosition = get(this, 'position');
+    }
+
+    let preserveScrollPosition = (get(transition, 'router.currentRouteInfos') || []).some((routeInfo) => get(routeInfo, 'route.controller.preserveScrollPosition'));
+
+    // If `preserveScrollPosition` was not set on the controller, attempt fallback to `preserveScrollPosition` which was set on the router service.
+    if(!preserveScrollPosition) {
```
Comment (jleja) [inline addon/services/router-scroll.js:133]:
```suggestion
    if (!preserveScrollPosition) {
```

File: addon/services/router-scroll.js [hunk 2]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+      preserveScrollPosition = get(this, 'preserveScrollPosition')
+    }
+
+    if (!preserveScrollPosition) {
```
Comment (jleja) [inline addon/services/router-scroll.js:137]: Can combine the two `if (!preserveScrollPosition)` blocks

File: addon/services/router-scroll.js [hunk 3]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+
+    const scrollWhenIdle = get(this, 'scrollWhenIdle');
+    const scrollWhenAfterRender = get(this, 'scrollWhenAfterRender');
+
+    if (!scrollWhenIdle && !scrollWhenAfterRender) {
+      // out of the option, this happens on the tightest schedule
+      scheduleOnce('render', this, CALLBACK, transition);
+    } else if (scrollWhenAfterRender && !scrollWhenIdle) {
+      // out of the option, this happens on the tightest schedule
+      scheduleOnce('afterRender', this, CALLBACK, transition);
```
Comment (jleja) [inline addon/services/router-scroll.js:182]: Are the duplicated comments here correct, or is scheduling after `render` the option that happens on the tightest schedule?

**Review by jleja — APPROVED** [review]
Looks good.  Just had two things to point out.

File: addon/services/router-scroll.js [hunk 4]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+    if (get(this, 'isFirstLoad')) {
+      this.unsetFirstLoad();
+    }
+
+    let scrollPosition;
+
+    if (url && url.indexOf('#') > -1 && hashElement) {
+      scrollPosition = { x: hashElement.offsetLeft, y: hashElement.offsetTop };
+    } else {
+      scrollPosition = get(this, 'position');
```
Comment (cbroeren) [inline addon/services/router-scroll.js:127]: What if you use this as the initial value (it gets overwritten in the if block if needed)

File: addon/services/router-scroll.js [hunk 5]
```diff
@@ -1,10 +1,61 @@
...
+ * @void
+ */
+function tryScrollRecursively(fn, scrollHash, element) {
+  let documentHeight;
+  // read DOM outside of rAF
+  if (element) {
+    documentHeight = Math.max(element.scrollHeight, element.offsetHeight, element.clientHeight);
+  } else {
+    const body = document.body;
+    const html = document.documentElement;
```
Comment (cbroeren) [inline addon/services/router-scroll.js:34]:
```suggestion
    const { body, documentElement: html } = document;
```

File: addon/services/router-scroll.js [hunk 6]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+    let preserveScrollPosition = (get(transition, 'router.currentRouteInfos') || []).some((routeInfo) => get(routeInfo, 'route.controller.preserveScrollPosition'));
+
+    // If `preserveScrollPosition` was not set on the controller, attempt fallback to `preserveScrollPosition` which was set on the router service.
+    if (!preserveScrollPosition) {
+      preserveScrollPosition = get(this, 'preserveScrollPosition')
+    }
+
+    if (!preserveScrollPosition) {
+      const scrollElement = get(this, 'scrollElement');
+      const targetElement = get(this, 'targetElement');
```
Comment (cbroeren) [inline addon/services/router-scroll.js:139]: Do you need the `get` or could you use `const { scrollElement, targetElement } = this;` ?

File: addon/services/router-scroll.js [hunk 7]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+    }
+
+    const scrollWhenIdle = get(this, 'scrollWhenIdle');
+    const scrollWhenAfterRender = get(this, 'scrollWhenAfterRender');
+
+    if (!scrollWhenIdle && !scrollWhenAfterRender) {
+      // out of the option, this happens on the tightest schedule
+      scheduleOnce('render', this, CALLBACK, transition);
+    } else if (scrollWhenAfterRender && !scrollWhenIdle) {
+      // out of the option, this happens on the tightest schedule
```
Comment (cbroeren) [inline addon/services/router-scroll.js:181]: Should this be the same comment as above?

**Review by cbroeren — COMMENTED** [review]
I wasn't able to verify this addon with a running app yet, but left some comments. Looks 👍  overall

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,9 @@
 
 - Updating to th v3.0 series was due to removing `scrollWhenPainted` as a config option.  Also, we fixed some hidden bugs with scheduling when to scroll to your last y position.
 
+## 4.0.0
+* [#270](https://github.com/DockYard/ember-router-scroll/pull/270) [Major]: use`RouterService` instead of Ember `Router`
```
[the remark this answers is not in the document]
Comment (snewcomer) [reply CHANGELOG.md:6]: Great idea!  I updated the README.

File: addon/services/router-scroll.js [hunk 9]
```diff
@@ -29,6 +80,111 @@ class RouterScroll extends Service {
...
+      scrollPosition = { x: hashElement.offsetLeft, y: hashElement.offsetTop };
+    } else {
+      scrollPosition = get(this, 'position');
+    }
+
+    let preserveScrollPosition = (get(transition, 'router.currentRouteInfos') || []).some((routeInfo) => get(routeInfo, 'route.controller.preserveScrollPosition'));
+
+    // If `preserveScrollPosition` was not set on the controller, attempt fallback to `preserveScrollPosition` which was set on the router service.
+    if (!preserveScrollPosition) {
+      preserveScrollPosition = get(this, 'preserveScrollPosition')
```
[the remark this answers is not in the document]
Comment (snewcomer) [reply addon/services/router-scroll.js:134]: That is a good idea!

File: addon/services/router-scroll.js (hunk 4 above)
Comment (snewcomer) [reply addon/services/router-scroll.js]: Another good find.  Thank you Chantal!

**Pull request merged**